pl_bolts/optimizers/lr_scheduler.py

```diff
@@ -0,0 +1,81 @@
+import math
+import warnings
+
+from torch.optim.lr_scheduler import _LRScheduler
+
+
+class LinearWarmupCosineAnnealingLR(_LRScheduler):
+    """
+    Sets the learning rate of each parameter group to follow a linear warmup schedule
+    between warmup_start_lr and base_lr followed by a cosine annealing schedule between
+    base_lr and eta_min.
+
+    .. warning::
+        It is recommended to call :func:`.step()` for :class:`LinearWarmupCosineAnnealingLR`
+        after each iteration as calling it after each epoch will keep the starting lr at
+        warmup_start_lr for the first epoch which is 0 in most cases.
+    
+    TODO: what to pass per step and how to use it in lightning
+
+    Args:
+        optimizer (Optimizer): Wrapped optimizer.
+        warmup_epochs (int): Maximum number of iterations for linear warmup
+        max_epochs (int): Maximum number of iterations
+        warmup_start_lr (float): Learning rate to start the linear warmup. Default: 0.
+        eta_min (float): Minimum learning rate. Default: 0.
+        last_epoch (int): The index of last epoch. Default: -1.
+
+    Example:
+        >>> fill out with lightning example
+    """
+
+    def __init__(
+        self,
+        optimizer,
+        warmup_epochs,
+        max_epochs,
+        warmup_start_lr=0.,
+        eta_min=0.,
+        last_epoch=-1
+    ):
+
+        self.warmup_epochs = warmup_epochs
+        self.max_epochs = max_epochs
+        self.warmup_start_lr = warmup_start_lr
+        self.eta_min = eta_min
+
+        super(LinearWarmupCosineAnnealingLR, self).__init__(optimizer, last_epoch)
+
+    def get_lr(self):
+        if not self._get_lr_called_within_step:
+            warnings.warn("To get the last learning rate computed by the scheduler, "
+                          "please use `get_last_lr()`.", UserWarning)
+
+        if self.last_epoch == 0:
+            return [self.warmup_start_lr] * len(self.base_lrs)
+        elif self.last_epoch < self.warmup_epochs:
+            return [group['lr'] + (base_lr - self.warmup_start_lr) / (self.warmup_epochs - 1)
+                    for base_lr, group in zip(self.base_lrs, self.optimizer.param_groups)]
+        elif self.last_epoch == self.warmup_epochs:
+            return self.base_lrs
+        # TODO: what does this do? This might be wrong for now since axis has shifted due to warmup
+        elif (self.last_epoch - 1 - self.max_epochs) % (2 * (self.max_epochs - self.warmup_epochs)) == 0:
```

Comment:
@williamFalcon @Borda this is the T_curr = 2 (k + 1) T_max case for linear warmup + cosine annealing scheduler. This should fail but it doesn't. Only when you adjust it for the linear warmup part it fails. This elif condition is not adjusted for the linear warmup case.